cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,120 @@
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        return describeMetadataQuorum().map(info -> {
+            boolean canRoll = isQuorumHealthyWithoutPod(podId, info);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        return describeMetadataQuorum().map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthyWithoutPod(int podId, QuorumInfo info) {
+        int leaderId = info.leaderId();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+        Map<Integer, Long> controllerStates = info.voters().stream().collect(Collectors.toMap(
+                QuorumInfo.ReplicaState::replicaId,
+                state -> state.lastCaughtUpTimestamp().isPresent() ? state.lastCaughtUpTimestamp().getAsLong() : -1));
+        int totalNumOfControllers = controllerStates.size();
+        if (totalNumOfControllers == 1) {
+            LOGGER.warnCr(reconciliation, "Performing rolling update on a controller quorum with a single node. The cluster may be " +
+                    "in a defective state once the rolling update is complete. It is recommended that a minimum of three controllers are used.");
+            return true;
+        }
+        //cannot use normal integer as it's being incremented inside the lambda expression
+        AtomicInteger numOfCaughtUpControllers = new AtomicInteger();
+        long leaderLastCaughtUpTimestamp = controllerStates.get(leaderId);
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        controllerStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
```

Comment:
here and above we are using "leader replica" or just "replica". We are talking about controller leader and controller (follower) replicas right? Can we make the log clearer about it?
It will also match with the naming we are using in the next logging lines, where "controller" is mentioned clearly.

Reply:
We'll update to be more consistent